### PR TITLE
[SPARK-2883][SQL] Orc support through datasource api

### PR DESCRIPTION
**Pull request by scwf**
Adding support for read/write orc files through the new datasource api.

### CODE REVIEW

**Comment by scwf** [comment]
@liancheng and @marmbrus , i removed the partitioned support for orc tables and added write interface based on the newly introduced write api, can you help review this? thanks

**Comment by krise3k** [comment]
+1

**Comment by scwf** [comment]
Retest this please

**Comment by scwf** [comment]
Retest this please.

File: core/src/main/scala/org/apache/spark/SparkHadoopWriter.scala [hunk 1]
```diff
@@ -41,20 +41,20 @@ class SparkHadoopWriter(@transient jobConf: JobConf)
...
-  @transient private var writer: RecordWriter[AnyRef,AnyRef] = null
-  @transient private var format: OutputFormat[AnyRef,AnyRef] = null
-  @transient private var committer: OutputCommitter = null
-  @transient private var jobContext: JobContext = null
-  @transient private var taskContext: TaskAttemptContext = null
+  @transient protected var writer: RecordWriter[AnyRef,AnyRef] = null
+  @transient protected var format: OutputFormat[AnyRef,AnyRef] = null
+  @transient protected var committer: OutputCommitter = null
+  @transient protected var jobContext: JobContext = null
+  @transient protected var taskContext: TaskAttemptContext = null
```
Comment (scwf) [inline core/src/main/scala/org/apache/spark/SparkHadoopWriter.scala]: I changed the scope of var/def of SparkHadoopWriter to reuse these code in orc writting api implementation

**Comment by scwf** [comment]
To make ORC datasource clean and easy to review, i will split it to three part of work, each one should be a PR.
1 orc datasource api support including read/write implementation, no partitioned support
2 filter push down optimization
3 partitioning support

This is the PR for the first point.

/cc @marmbrus @liancheng

**Comment by scwf** [comment]
ping

**Comment by scwf** [comment]
i am closing this in favor of #6914

**Pull request closed without merging**